File: AN_2/SEM_2/MPP/c#/lab4/lab4/repository/ParticipantRepository.cs
using System.Collections;

namespace lab4.repository;
using lab4.utils;
using lab4.domain;
using log4net;
using Microsoft.Data.Sqlite;

public class ParticipantRepository : IParticipantRepository
{
    private static readonly ILog Logger = LogManager.GetLogger(typeof(ParticipantRepository));
    private readonly DatabaseUtils _dbUtils;

    public ParticipantRepository(string connectionString)
    {
        _dbUtils = new DatabaseUtils();
    }

    public void Adauga(Participant participant)
    {
        Logger.Debug($"Incep adaugarea participantului: {participant.GetNume()}");
        string sql =
            "INSERT INTO participant (nume, CNP, capacitate_motor, idEchipa) VALUES (@nume, @CNP, @capacitate_motor, @idEchipa)";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@nume", participant.GetNume());
                command.Parameters.AddWithValue("@CNP", participant.GetCnp());
                command.Parameters.AddWithValue("@capacitate_motor", participant.GetCapacitateMotor());
                command.Parameters.AddWithValue("@idEchipa", participant.GetEchipa().GetId());

                connection.Open();
                command.ExecuteNonQuery();

                var getIdCommand = new SqliteCommand("SELECT last_insert_rowid();", connection);
                var lastInsertRowId = (long)getIdCommand.ExecuteScalar();
                participant.SetId((int)lastInsertRowId);

                Logger.Debug($"Participantul {participant.GetNume()} a fost adaugat cu succes.");
            }
        }
        catch (Exception e)
        {
            Logger.Error($"Eroare la adaugarea participantului: {participant.GetNume()}", e);
        }
    }

    public void Sterge(int id)
    {
        Logger.Debug($"Incep stergerea participantului cu id {id}");
        string sql = "DELETE FROM participant WHERE idParticipant = @idParticipant";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idParticipant", id);

                connection.Open();
                command.ExecuteNonQuery();
                Logger.Debug($"Participantul cu id {id} a fost sters cu succes.");
            }
        }
        catch (Exception e)
        {
            Logger.Error($"Eroare la stergerea participantului cu id {id}", e);
        }
    }

    public void Modifica(Participant participant)
    {
        Logger.Debug($"Incep modificarea participantului cu id {participant.GetId()}");
        string sql =
            "UPDATE participant SET nume = @nume, CNP = @CNP, capacitate_motor = @capacitate_motor, idEchipa = @idEchipa WHERE idParticipant = @idParticipant";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@nume", participant.GetNume());
                command.Parameters.AddWithValue("@CNP", participant.GetCnp());
                command.Parameters.AddWithValue("@capacitate_motor", participant.GetCapacitateMotor());
                command.Parameters.AddWithValue("@idEchipa", participant.GetEchipa().GetId());
                command.Parameters.AddWithValue("@idParticipant", participant.GetId());

                connection.Open();
                command.ExecuteNonQuery();

                Logger.Debug($"Participantul cu id {participant.GetId()} a fost modificat cu succes.");
            }
        }
        catch (Exception e)
        {
            Logger.Error($"Eroare la modificarea participantului cu id {participant.GetId()}", e);
        }
    }


    public Participant FindById(int id)
    {
        Logger.Debug($"Incep cautarea participantului cu id {id}");
        string sql = @"
        SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, 
               e.idEchipa, e.nume AS echipaNume
        FROM participant p
        JOIN echipa e ON p.idEchipa = e.idEchipa
        WHERE p.idParticipant = @id";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@id", id);
                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        int participantId = reader.GetInt32(0);
                        string nume = reader.GetString(1);
                        string cnp = reader.GetString(2);
                        int capacitateMotor = reader.GetInt32(3);
                        int echipaId = reader.GetInt32(4);
                        string echipaNume = reader.GetString(5);

                        Echipa echipa = new Echipa(echipaId, echipaNume);
                        Participant participant = new Participant(participantId, nume, cnp, capacitateMotor, echipa);

                        Logger.Debug($"Participantul cu id {id} a fost gasit.");
                        return participant;
                    }
                }
            }
        }
        catch (Exception e)
        {
            Logger.Error($"Eroare la cautarea participantului cu id {id}", e);
        }

        Logger.Debug($"Participantul cu id {id} nu a fost gasit.");
        return null;
    }

    public List<Participant> FindAll()
    {
        Logger.Debug("Incep cautarea tuturor participantilor.");
        List<Participant> participants = new List<Participant>();
        string sql = @"
        SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, 
               e.idEchipa, e.nume AS echipaNume
        FROM participant p
        JOIN echipa e ON p.idEchipa = e.idEchipa";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        int participantId = reader.GetInt32(0);
                        string nume = reader.GetString(1);
                        string cnp = reader.GetString(2);
                        int capacitateMotor = reader.GetInt32(3);
                        int echipaId = reader.GetInt32(4);
                        string echipaNume = reader.GetString(5);

                        Echipa echipa = new Echipa(echipaId, echipaNume);
                        participants.Add(new Participant(participantId, nume, cnp, capacitateMotor, echipa));
                    }
                }
            }

            Logger.Debug("Cautarea tuturor participantilor s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            Logger.Error("Eroare la preluarea tuturor participantilor.", e);
        }

        return participants;
    }


    public List<ParticipantDTO> FindByEchipa(Echipa echipa)
    {
        Logger.Debug($"Incep cautarea participantilor din echipa cu id {echipa.GetId()}");
        List<ParticipantDTO> result = new List<ParticipantDTO>();
        string sql = "SELECT * FROM participant WHERE idEchipa = @idEchipa";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@idEchipa", echipa.GetId());

                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        string nume = reader.GetString(reader.GetOrdinal("nume"));
                        int capacitateMotor = reader.GetInt32(reader.GetOrdinal("capacitate_motor"));
                        result.Add(new ParticipantDTO( nume, capacitateMotor, echipa.GetNume()));
                    }
                }
            }

            Logger.Debug($"Cautarea participantilor din echipa cu id {echipa.GetId()} s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            Logger.Error($"Eroare la cautarea participantilor din echipa cu id {echipa.GetId()}", e);
        }

        return result;
    }
    public List<Participant> FindByNume(string nume)
    {
        Logger.Debug($"Incep cautarea participantilor cu numele {nume}");
        List<Participant> result = new List<Participant>();
        string sql = "SELECT * FROM participant WHERE nume = @nume";

        try
        {
            using (var connection = _dbUtils.GetConnection())
            using (var command = new SqliteCommand(sql, connection))
            {
                command.Parameters.AddWithValue("@nume", nume);

                connection.Open();
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        int participantId = reader.GetInt32(reader.GetOrdinal("idParticipant"));
                        string numeParticipant = reader.GetString(reader.GetOrdinal("nume"));
                        string cnp = reader.GetString(reader.GetOrdinal("CNP"));
                        int capacitateMotor = reader.GetInt32(reader.GetOrdinal("capacitate_motor"));
                        int echipaId = reader.GetInt32(reader.GetOrdinal("idEchipa"));
                        string echipaNume = reader.GetString(reader.GetOrdinal("nume"));

                        Echipa echipa = new Echipa(echipaId, echipaNume);
                        result.Add(new Participant(participantId, numeParticipant, cnp, capacitateMotor, echipa));
                    }
                }
            }

            Logger.Debug($"Cautarea participantilor cu numele {nume} s-a realizat cu succes.");
        }
        catch (Exception e)
        {
            Logger.Error($"Eroare la cautarea participantilor cu numele {nume}.", e);
        }

        return result;
    }
    


}

File: AN_2/SEM_2/MPP/java/lab7/Persistence/src/main/java/ro/mpp2024/persistence/repository/ParticipantRepository.java
package ro.mpp2024.persistence.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ro.mpp2024.model.Echipa;
import ro.mpp2024.model.Participant;
import ro.mpp2024.persistence.ParticipantRepositoryInterface;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantRepository implements ParticipantRepositoryInterface {
    private final JdbcUtils _dbUtils;
    private static final Logger logger = LogManager.getLogger(ParticipantRepository.class);

    public ParticipantRepository(Properties props) {
        this._dbUtils = new JdbcUtils(props);
    }

    @Override
    public void adauga(Participant participant) {
        logger.traceEntry("Incep adaugarea participantului: " + participant.getNume());
        String sql = "INSERT INTO participant (nume, CNP, capacitate_motor, idEchipa) VALUES (?, ?, ?, ?)";
        try (Connection connection = _dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {
            statement.setString(1, participant.getNume());
            statement.setString(2, participant.getCNP());
            statement.setInt(3, participant.getCapacitateMotor());
            statement.setInt(4, participant.getEchipa().getId());
            statement.executeUpdate();
            logger.traceExit("Participantul " + participant.getNume() + " a fost adaugat cu succes.");
            try (ResultSet generatedKeys = statement.getGeneratedKeys()) {
                if (generatedKeys.next()) {
                    participant.setId(generatedKeys.getInt(1));
                }
            }
        } catch (SQLException e) {
            logger.error("Eroare la adaugarea participantului: " + participant.getNume(), e);
        }
    }

    @Override
    public void sterge(Integer id) {
        logger.traceEntry("Incep stergerea participantului cu id " + id);
        String sql = "DELETE FROM participant WHERE idParticipant = ?";
        try (Connection connection = _dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setInt(1, id);
            statement.executeUpdate();
            logger.traceExit("Participantul cu id " + id + " a fost sters cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la stergerea participantului cu id " + id, e);
        }
    }

    @Override
    public void modifica(Participant participant) {
        logger.traceEntry("Incep modificarea participantului cu idParticipant " + participant.getId());
        String sql = "UPDATE participant SET nume = ?, CNP = ?, capacitate_motor = ?, idEchipa = ? WHERE idParticipant = ?";
        try (Connection connection = _dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, participant.getNume());
            statement.setString(2, participant.getCNP());
            statement.setInt(3, participant.getCapacitateMotor());
            statement.setInt(4, participant.getEchipa().getId());
            statement.setInt(5, participant.getId());
            statement.executeUpdate();
            logger.traceExit("Participantul cu idParticipant " + participant.getId() + " a fost modificat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la modificarea participantului cu id " + participant.getId(), e);
        }
    }

    @Override
    public Participant findById(Integer id) {
        logger.traceEntry("Incep cautarea participantului cu id " + id);
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume " +
                "FROM participant p " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa " +
                "WHERE p.idParticipant = ?";
        try (Connection connection = _dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setInt(1, id);
            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                Participant participant = new Participant(
                        resultSet.getInt("idParticipant"),
                        resultSet.getString("nume"),
                        resultSet.getString("CNP"),
                        resultSet.getInt("capacitate_motor"),
                        new Echipa(resultSet.getInt("idEchipa"), resultSet.getString("echipaNume"))
                );
                logger.traceExit("Participantul cu id " + id + " a fost gasit.");
                return participant;
            }
        } catch (SQLException e) {
            logger.error("Eroare la cautarea participantului cu id " + id, e);
        }
        logger.traceExit("Participantul cu id " + id + " nu a fost gasit.");
        return null;
    }

    @Override
    public List<Participant> findAll() {
        logger.traceEntry("Incep cautarea tuturor participantilor.");
        List<Participant> participantList = new ArrayList<>();
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume " +
                "FROM participant p " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa";
        try (Connection connection = _dbUtils.getConnection();
             Statement statement = connection.createStatement();
             ResultSet resultSet = statement.executeQuery(sql)) {
            while (resultSet.next()) {
                Participant participant = new Participant(
                        resultSet.getInt("idParticipant"),
                        resultSet.getString("nume"),
                        resultSet.getString("CNP"),
                        resultSet.getInt("capacitate_motor"),
                        new Echipa(resultSet.getInt("idEchipa"), resultSet.getString("echipaNume"))
                );
                participantList.add(participant);
            }
            logger.traceExit("Cautarea tuturor participantilor s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la preluarea tuturor participantilor.", e);
        }
        return participantList;
    }

    @Override
    public List<Participant> findByEchipa(Echipa echipa) {
        logger.traceEntry("Incep cautarea participantilor din echipa cu id " + echipa.getId());
        List<Participant> result = new ArrayList<>();
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume " +
                "FROM participant p " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa " +
                "WHERE p.idEchipa = ?";
        try (Connection connection = _dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setInt(1, echipa.getId());
            ResultSet resultSet = statement.executeQuery();
            while (resultSet.next()) {
                Echipa ech = new Echipa(resultSet.getInt("idEchipa"), resultSet.getString("echipaNume"));
                Participant participant = new Participant(
                        resultSet.getInt("idParticipant"),
                        resultSet.getString("nume"),
                        resultSet.getString("CNP"),
                        resultSet.getInt("capacitate_motor"),
                        ech
                );
                result.add(participant);
            }
            logger.traceExit("Cautarea participantilor din echipa cu id " + echipa.getId() + " s-a realizat cu succes.");
        } catch (SQLException e) {
            logger.error("Eroare la cautarea participantilor din echipa cu id " + echipa.getId(), e);
        }
        return result;
    }


    @Override
    public Participant findByNume(String nume) {
        logger.traceEntry("Incep cautarea participantului cu numele " + nume);
        String sql = "SELECT p.idParticipant, p.nume, p.CNP, p.capacitate_motor, e.idEchipa, e.nume as echipaNume " +
                "FROM participant p " +
                "JOIN echipa e ON p.idEchipa = e.idEchipa " +
                "WHERE p.nume = ?";
        try (Connection connection = _dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setString(1, nume);
            ResultSet resultSet = statement.executeQuery();
            if (resultSet.next()) {
                Participant participant = new Participant(
                        resultSet.getInt("idParticipant"),
                        resultSet.getString("nume"),
                        resultSet.getString("CNP"),
                        resultSet.getInt("capacitate_motor"),
                        new Echipa(resultSet.getInt("idEchipa"), resultSet.getString("echipaNume"))
                );
                logger.traceExit("Participantul cu numele " + nume + " a fost gasit.");
                return participant;
            }
        } catch (SQLException e) {
            logger.error("Eroare la cautarea participantului cu numele " + nume, e);
        }
        logger.traceExit("Participantul cu numele " + nume + " nu a fost gasit.");
        return null;
    }


}
